### PR TITLE
[big boy] UI Init

### PR DESCRIPTION
**Pull request by chriskrogh**
Worth a read

## UI
Screen-Recording-2020-06-03-at-8

## What it took to get there
1. I started off by adding 3 new dependencies for our UI
* [MUI Core](https://material-ui.com/) - Contains everything we need for theming and styling our components.
Also contains our modular input components
* [MUI Icons](https://www.npmjs.com/package/@material-ui/icons) - The reason why I use this over FA or some other lighter icon packages is because of how well it integrates with MUI theming.
* [classnames](https://www.npmjs.com/package/classnames) - **super light package** - used for combining class names 

2. Then I tackled 2 components `Page` and `ToggleThemeButton`

Essentially how we're handling theming is using `ThemeProvider` by MUI to provide **A** theme to our app and we're using [React Context](https://reactjs.org/docs/context.html) to create `ThemeContextProvider` to specify **which** theme gets used by the `ThemeProvider`.

Right now the 2 themes are `light` and `dark` and the color schemes are arbitrary. (they should be updated)

### CODE REVIEW

File: client/src/components/Page/index.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { withStyles, createStyles } from '@material-ui/core/styles';
+import /*type*/ { WithStyles, Theme } from '@material-ui/core/styles';
```
Comment (jtaaa) [inline client/src/components/Page/index.tsx:3]: What's up with this `/*type*/`?

File: client/src/components/Page/index.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+function Page({ classes, children }: Props) {
+  return (
+    <div className={classes.page}>
+      {children}
+    </div>
+  );
+}
+
+export default withStyles(styles, { withTheme: true })(Page);
```
Comment (jtaaa) [inline client/src/components/Page/index.tsx:30]:
```suggestion
export default withStyles(styles, { withTheme: true })(Page);

```

File: client/src/components/buttons/ThemeToggle/index.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        <Brightness4Icon className={classes.iconSize} />
+      </div>
+      <div className={classes.labelContainer}>
+        <p>Toggle Theme</p>
+      </div>
+    </Button>
+  );
+}
+
+export default withStyles(styles)(ThemeToggleButton);
```
Comment (jtaaa) [inline client/src/components/buttons/ThemeToggle/index.tsx:58]:
```suggestion
export default withStyles(styles)(ThemeToggleButton);

```

File: client/src/context/theme/actions.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+type SetTheme = {
+  type: ActionTypes.SET_THEME;
+  payload: NamedTheme;
+}
+
+export type Action = SetTheme;
+
+export default ActionTypes;
```
Comment (jtaaa) [inline client/src/context/theme/actions.ts:14]:
```suggestion
export default ActionTypes;

```

File: client/src/context/theme/provider.tsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <ThemeContext.Provider value={{
+      ...state,
+      setTheme,
+    }}>
+      {children}
+    </ThemeContext.Provider>
+  );
+}
+
+export default Provider;
```
Comment (jtaaa) [inline client/src/context/theme/provider.tsx:29]:
```suggestion
export default Provider;

```

File: client/src/context/theme/reducer.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  switch (action.type) {
+    case ActionTypes.SET_THEME:
+      return {
+        ...state,
+        theme: action.payload
+      };
+    default:
+      return state;
+  }
+}
```
Comment (jtaaa) [inline client/src/context/theme/reducer.ts:15]:
```suggestion
}

```

File: client/src/context/theme/state.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+import LightTheme from 'theme/themes/light';
+import DarkTheme from 'theme/themes/dark';
+
+export const initialState = {
+  theme: (localStorage.getItem('theme') === 'dark' ? DarkTheme : LightTheme),
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  setTheme: (theme: NamedTheme) => { },
+};
+
+export const ThemeContext = createContext(initialState);
```
Comment (jtaaa) [inline client/src/context/theme/state.ts:12]:
```suggestion
export const ThemeContext = createContext(initialState);

```

File: client/src/components/buttons/ThemeToggle/index.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    justifyContent: 'center',
+    alignItems: 'center',
+  }
+});
+
+function ThemeToggleButton({ classes }: WithStyles<typeof styles>) {
+  const { theme, setTheme } = useContext(ThemeContext);
+
+  const toggleTheme = () => {
+    setTheme(theme.name === 'light' ? DarkTheme : LightTheme);
```
Comment (jtaaa) [inline client/src/components/buttons/ThemeToggle/index.tsx:43]:
```suggestion
    setTheme(theme.name === LightTheme.name ? DarkTheme : LightTheme);
```

File: client/src/components/Page/index.tsx (hunk 1 above)
Comment (chriskrogh) [reply client/src/components/Page/index.tsx:3]: So there's this new syntax that CRA doesn't have available yet. [link](https://stackoverflow.com/questions/59833120/parsing-error-expected-in-import-type)

When the syntax becomes available, I want to use it

Essentially I like to split `type` imports  from other imports and have them closer to the top of the import list 
Makes TS more readable I guess

Comment (jtaaa) [reply client/src/components/Page/index.tsx:3]: Ooo ok cool

**Comment by chriskrogh** [comment]
Damn I should have responded to your other comment before merging this.

This is an example of how I do prop Typing in TS

```
interface Props extends WithStyles<typeof styles> {
  banner?: string;
  picture?: string;
  panel: Panels.PROFILE | Panels.COMMUNITY;
}

function BannerPicture({ classes, banner, picture, panel }: Props) {
```

We can use this PR to continue the discussion @jtaaa

**Comment by jtaaa** [comment]
I'll follow suit 👍

**Pull request merged**